Simplify attribute interpolations

crisbeto opened the pull request:
Replaces the existing attribute interpolation instructions with a new format. For example previously we would generate `attributeInterpolate(name, ...)` whereas now we generate `attribute(name, interpolate(...))`. This allows us to remove all of the specialized interpolation instructions.

This initial PR only deals with attributes, but if there aren't any breakages, we can roll out the approach to all interpolations, saving us ~40 instructions from the set.

crisbeto committed:
refactor(core): add instructions for interpolating values

Adds the new `interpolate*` instructions that can be passed into other instructions and used to replace our existing flavors of interpolations.

crisbeto committed:
refactor(compiler): replace attribute interpolation instructions

Replaces the attribute interpolation instructions with `attribute` plus the new `interpolateX` instruction. This allows to reduce our overall instruction footprint.

crisbeto committed:
refactor(core): remove attribute interpolation instructions

The attribute interpolation instructions aren't used anymore so we can remove them.

crisbeto commented on packages/core/src/render3/instructions/attribute.ts:
```diff
@@ -29,13 +30,18 @@ export function ɵɵattribute(
-    ngDevMode && storePropertyBindingMetadata(tView.data, tNode, 'attr.' + name, bindingIndex);
+
+  // Value can be `NO_CHANGE` in case of an interpolation.
+  if (value !== NO_CHANGE) {
```
The issue here is that `bindingUpdated` has an assertion against `NO_CHANGE`, however it also seems a bit wasteful to introduce another instruction purely because of this one (inexpensive) check.

pkozlowski-opensource replied:
An alternative might be to drop the check in the `bindingUpdated` and replace it with the `(value !== NO_CHANGE)` guard

crisbeto commented on packages/core/test/acceptance/change_detection_spec.ts:
```diff
@@ -1671,7 +1671,7 @@ describe('change detection', () => {
 
     it('should include field name in case of attribute interpolation', () => {
-      const message = `Previous value for 'attr.id': 'Expressions: a and initial!'. Current value: 'Expressions: a and changed!'`;
+      const message = `Expression has changed after it was checked. Previous value: 'initial'. Current value: 'changed'`;
```
This changed, because the error comes from the `interpolate` instructions which don't know what attribute they're bound to.

pkozlowski-opensource replied:
Right, this makes the error message less useful but I think it is the tradeoff we should make

crisbeto commented:
[Passing TGP](http://test/OCL:761457164:BASE:761511210:1747836523801:125b651d)

crisbeto committed:
refactor(core): handle NO_CHANGE in bindingUpdated

Updates the `bindingUpdated` function to handle `NO_CHANGE` instead of throwing. This will allow us to reuse instructions across more cases.

thePunderWoman commented:
This PR was merged into the repository by commit 1650a8572a3635511cd294265b347afb177b8812.

The changes were merged into the following branches: main, 20.0.x
